File: Week_02/144_binary_tree_preorder_traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Recursion
class Solution0 {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();

        helper(root, res);

        return res;
    }

    private void helper(TreeNode root, List<Integer> result){
        if(root == null) return;
        result.add(root.val);
        if(root.left != null) helper(root.left, result);
        if(root.right != null) helper(root.right, result);
    }
}


// Stack
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        if(root == null) return res;

        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode node = root;

        while(!stack.isEmpty() || node != null) {
            while(node != null) {
                res.add(node.val);
                stack.push(node);
                node = node.left;
            }
            node = stack.pop();
            node = node.right;
        }

        return res;
    }
}

File: Week_03/python/tree_traverse/144_binary_tree_preorder_traversal.py
# Definition for a binary tree node.
from collections import deque
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        return self.preorder(root)

    def preorder(self, root: TreeNode):
        if not root: return []
        return [root.val] + self.preorder(root.left) + self.preorder(root.right)


class Solution0:
    def preorderTraversal(self, root: TreeNode) -> List[int]:
        stack = deque()
        res = []

        node = root

        while len(stack) > 0 or node:
            while node:
                res.append(node.val)
                stack.append(node)
                node = node.left

            node = stack.pop()
            node = node.right

        return res
